File: Week 30/python/top_k_elements.py
"""
WEEK 30 - PYTHON ADVANCED TOPICS
Topic: Top-K Elements Pattern
File: top_k_elements.py

CONCEPT:
    Many interview problems reduce to "find the K largest / smallest /
    most-frequent elements". Three idiomatic solutions:
      1. Min-heap of size K (keeps largest): O(n log K), O(K) space.
      2. Bucket sort by frequency: O(n) time, O(n) space (when input is
         integer or hashable with bounded counts).
      3. Quickselect partition: O(n) expected time, O(1) extra space.

KEY POINTS:
    - Min-heap of size K is the safe default when K << n.
    - Bucket sort works for top-K frequent because counts are bounded by n.
    - Quickselect (Hoare partitioning) is the fastest but has O(n^2) worst
      case unless you randomise or use median-of-medians.

ALGORITHM / APPROACH:
    KTH LARGEST (heap):
        push x into heap; if size > K: pop. top = K-th largest.
    TOP K FREQUENT (heap):
        count freq; min-heap on freq with size K.
    TOP K FREQUENT (bucket):
        count freq; buckets[freq].append(item); collect from highest bucket.
    QUICKSELECT:
        pivot partition; recurse into the side containing K.

PYTHON-SPECIFIC NOTES:
    - `heapq` is a min-heap; for max-heap behaviour negate keys.
    - `collections.Counter` for frequency tally.

DRY RUN / EXAMPLE:
    find_kth_largest [3,2,1,5,6,4], k=2 -> 5
    top_k_frequent  [1,1,1,2,2,3], k=2 -> [1, 2]
    quickselect_kth [3,2,1,5,6,4], k=2 -> 5

COMPLEXITY:
    Heap-based:    O(n log K) time, O(K) space
    Bucket:        O(n) time, O(n) space
    Quickselect:   O(n) expected time, O(1) extra space
"""

from __future__ import annotations

import heapq
import random
from collections import Counter
from typing import List


def find_kth_largest(nums: List[int], k: int) -> int:
    heap: List[int] = []
    for x in nums:
        heapq.heappush(heap, x)
        if len(heap) > k:
            heapq.heappop(heap)
    return heap[0]


def top_k_frequent_heap(nums: List[int], k: int) -> List[int]:
    freq = Counter(nums)
    heap: List[tuple] = []
    for v, c in freq.items():
        heapq.heappush(heap, (c, v))
        if len(heap) > k:
            heapq.heappop(heap)
    return [v for _, v in heap]


def top_k_frequent_bucket(nums: List[int], k: int) -> List[int]:
    freq = Counter(nums)
    buckets: List[List[int]] = [[] for _ in range(len(nums) + 1)]
    for v, c in freq.items():
        buckets[c].append(v)
    out: List[int] = []
    for i in range(len(buckets) - 1, -1, -1):
        for v in buckets[i]:
            out.append(v)
            if len(out) == k:
                return out
    return out


def quickselect_kth_largest(nums: List[int], k: int) -> int:
    """Hoare-partition quickselect (expected O(n))."""
    nums = nums[:]  # don't mutate caller
    target = len(nums) - k

    def partition(lo: int, hi: int) -> int:
        pivot = nums[random.randint(lo, hi)]
        i, j = lo, hi
        while i <= j:
            while nums[i] < pivot: i += 1
            while nums[j] > pivot: j -= 1
            if i <= j:
                nums[i], nums[j] = nums[j], nums[i]
                i += 1; j -= 1
        return i

    lo, hi = 0, len(nums) - 1
    while lo < hi:
        idx = partition(lo, hi)
        if idx <= target:
            lo = idx
        else:
            hi = idx - 1
    return nums[target]


def _demo() -> None:
    print(f"Kth largest [3,2,1,5,6,4] k=2: {find_kth_largest([3,2,1,5,6,4], 2)}")
    print(f"Top 2 frequent (heap):   {sorted(top_k_frequent_heap([1,1,1,2,2,3], 2))}")
    print(f"Top 2 frequent (bucket): {sorted(top_k_frequent_bucket([1,1,1,2,2,3], 2))}")
    print(f"Quickselect kth=2: {quickselect_kth_largest([3,2,1,5,6,4], 2)}")


if __name__ == "__main__":
    _demo()


# NOTES
# -----
# Differences from Java:
#   * heapq is a min-heap; use tuples for composite priorities.
#   * Counter / Counter.most_common(k) are sometimes simpler than the
#     manual heap / bucket approaches; we keep both for didactic value.

File: Week 30/java/top_k_elements.java
/*
 * WEEK 30 - JAVA ADVANCED TOPICS
 * Topic: Top-K Elements Pattern
 * File: top_k_elements.java
 *
 * CONCEPT:
 *     Many interview problems reduce to "find the K largest / smallest /
 *     most-frequent elements". Three idiomatic solutions:
 *       1. Min-heap of size K (keeps largest): O(n log K), O(K) space.
 *       2. Bucket sort by frequency: O(n) time, O(n) space (when input is
 *          integer / hashable with bounded counts).
 *       3. Quickselect partition: O(n) expected time, O(1) extra space.
 *
 * KEY POINTS:
 *     - Min-heap of size K is the safe default when K << n.
 *     - Bucket sort works for top-K frequent because counts are bounded by n.
 *     - Quickselect (Hoare partition) is the fastest but has O(n^2) worst
 *       case unless you randomise or use median-of-medians.
 *
 * ALGORITHM / APPROACH:
 *     KTH LARGEST (heap):
 *         push x into heap; if size > K: pop. top = K-th largest.
 *     TOP K FREQUENT (heap):
 *         count freq; min-heap on freq with size K.
 *     TOP K FREQUENT (bucket):
 *         count freq; buckets[freq].append(item); collect from highest bucket.
 *     QUICKSELECT:
 *         pivot partition; recurse into the side containing K.
 *
 * DRY RUN / EXAMPLE:
 *     findKthLargest [3,2,1,5,6,4], k=2 -> 5
 *     topKFrequent  [1,1,1,2,2,3], k=2 -> [1, 2]
 *     quickselect    [3,2,1,5,6,4], k=2 -> 5
 *
 * COMPLEXITY:
 *     Heap-based:  O(n log K) time, O(K) space.
 *     Bucket:      O(n) time, O(n) space.
 *     Quickselect: O(n) expected time, O(1) extra space.
 */

// snake_case filename is fine; class TopKElements is package-private.

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Random;

class TopKElements {

    static int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        for (int x : nums) {
            heap.offer(x);
            if (heap.size() > k) heap.poll();
        }
        return heap.peek();
    }

    static int[] topKFrequentHeap(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();
        for (int x : nums) freq.merge(x, 1, Integer::sum);
        PriorityQueue<Map.Entry<Integer, Integer>> heap =
            new PriorityQueue<>(Comparator.comparingInt(Map.Entry::getValue));
        for (Map.Entry<Integer, Integer> e : freq.entrySet()) {
            heap.offer(e);
            if (heap.size() > k) heap.poll();
        }
        int[] result = new int[k];
        for (int i = 0; i < k; i++) result[i] = heap.poll().getKey();
        return result;
    }

    @SuppressWarnings("unchecked")
    static int[] topKFrequentBucket(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();
        for (int x : nums) freq.merge(x, 1, Integer::sum);
        List<Integer>[] buckets = new List[nums.length + 1];
        for (Map.Entry<Integer, Integer> e : freq.entrySet()) {
            int f = e.getValue();
            if (buckets[f] == null) buckets[f] = new ArrayList<>();
            buckets[f].add(e.getKey());
        }
        int[] out = new int[k];
        int idx = 0;
        for (int i = buckets.length - 1; i >= 0 && idx < k; i--) {
            if (buckets[i] == null) continue;
            for (int v : buckets[i]) {
                out[idx++] = v;
                if (idx == k) break;
            }
        }
        return out;
    }

    /** Hoare-partition quickselect (expected O(n)). Returns the K-th largest. */
    static int quickselectKthLargest(int[] nums, int k) {
        int[] arr = nums.clone();
        int target = arr.length - k;
        Random rng = new Random(0);
        int lo = 0, hi = arr.length - 1;
        while (lo < hi) {
            int pivot = arr[lo + rng.nextInt(hi - lo + 1)];
            int i = lo, j = hi;
            while (i <= j) {
                while (arr[i] < pivot) i++;
                while (arr[j] > pivot) j--;
                if (i <= j) {
                    int t = arr[i]; arr[i] = arr[j]; arr[j] = t;
                    i++; j--;
                }
            }
            if (i <= target) lo = i;
            else hi = i - 1;
        }
        return arr[target];
    }

    public static void main(String[] args) {
        System.out.println("Kth largest [3,2,1,5,6,4] k=2: "
            + findKthLargest(new int[]{3, 2, 1, 5, 6, 4}, 2));
        int[] heapResult = topKFrequentHeap(new int[]{1, 1, 1, 2, 2, 3}, 2);
        Arrays.sort(heapResult);
        System.out.println("Top 2 frequent (heap):   " + Arrays.toString(heapResult));
        int[] bucketResult = topKFrequentBucket(new int[]{1, 1, 1, 2, 2, 3}, 2);
        Arrays.sort(bucketResult);
        System.out.println("Top 2 frequent (bucket): " + Arrays.toString(bucketResult));
        System.out.println("Quickselect kth=2: "
            + quickselectKthLargest(new int[]{3, 2, 1, 5, 6, 4}, 2));
    }
}

/*
 * NOTES
 * -----
 * Differences from the Python implementation in top_k_elements.py:
 *   - Java's PriorityQueue is a min-heap by default; we pass an explicit
 *     Comparator for entries with composite keys.
 *   - The bucket array uses raw List[] with @SuppressWarnings; Java's
 *     generics-array limitations force this idiom.
 *   - Quickselect uses a seeded Random for deterministic output across runs.
 *   - The companion interview_patterns.java already covers findKthLargest /
 *     topKFrequent; we add quickselect for parity with the Python split.
 */
